### PR TITLE
Add Task to write a file onto a workspace

### PR DESCRIPTION
**Pull request by bittrance**
# Changes

This PR adds a catalog task to write a parameter out as a file, creating intermediate directories and setting permissions as necessary. While it is possible today to have a workspace that is a mounted config map variable, that provides limited support for expanding parameters into the file (i.e. the [replace-tokens](https://hub.tekton.dev/tekton/task/replace-tokens) task). Ultimately, my argument for adopting this task into the catalog is that I believe others will also look for a file-writing task and their experience of Tekton will be better if they find one.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

### CODE REVIEW

File: task/write-file/0.1/tests/run.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        name: write-file
+        kind: Task
+      workspaces:
+        - name: output
+          workspace: shared-workspace
+      params:
+        - name: path
+          value: ./config/login.ini
+        - name: permissions
+          value: "0400"
```
Comment (bittrance) [inline task/write-file/0.1/tests/run.yaml:20]: Waking up this morning, I realized that this should perhaps be `mode`? It seems Ansible, Puppet and Chef all call this parameter mode.

File: task/write-file/0.1/write-file.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      type: string
+      description: >
+        Relative path to create within the workspace. Directories will
+        be created as necessary.
+    - name: contents
+      type: string
+      description: >
+        Contents of the file to create. Note that octal numbers
+        need quoting in YAML.
+    - name: permissions
```
Comment (bittrance) [inline task/write-file/0.1/write-file.yaml:29]: Waking up this morning, I realized that this should perhaps be mode? It seems Ansible, Puppet and Chef all call this parameter mode.

File: task/write-file/0.1/write-file.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  steps:
+    - name: write-file
+      image: docker.io/library/alpine:3.10
+      workingDir: $(workspaces.output.path)
+      script: |
+        #!/bin/sh
+        set -e
+        DIRNAME=$(dirname "$(params.path)")
+        mkdir -p "$DIRNAME"
+        cat <<EOF > "./$(params.path)"
```
Comment (sbwsg) [inline task/write-file/0.1/write-file.yaml:47]: If possible, suggest passing through `printf` ala the example at https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#substituting-in-script-blocks

```yaml
# Task.yaml
spec:
  steps:
  - image: an-image-that-runs-bash
    env:
    - name: SCRIPT_CONTENTS
      value: $(params.script)
    script: |
      printf '%s' "${SCRIPT_CONTENTS}" > input-script
```

File: task/write-file/0.1/write-file.yaml (hunk 2 above)
Comment (bittrance) [reply task/write-file/0.1/write-file.yaml]: Fixed.

File: task/write-file/0.1/write-file.yaml (hunk 3 above)
Comment (bittrance) [reply task/write-file/0.1/write-file.yaml]: Fixed.

Comment (bittrance) [reply task/write-file/0.1/write-file.yaml]: Sorry for the flip-flopping; I was inspired by the style in #711 to change to `PARAM_.`

**Comment by sbwsg** [comment]
/lgtm

Thanks for this @bittrance !

**Comment by bittrance** [comment]
@sbwsg  Not familiar with the workflow here; there were several fixups that I now cleaned up. This seems to have provoked the bot to remove your lgtm label. Sorry about that. FWIW, I now consider the PR ready.

File: task/write-file/0.1/write-file.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      default: "0755"
+      description: >
+        chmod-style permission string to apply to the file. Note that
+        mode will not be applied to created directories.
+  workspaces:
+    - name: output
+      description: Workspace onto which the file is written.
+  steps:
+    - name: write-file
+      image: docker.io/library/alpine:3.10
```
Comment (vinamra28) [inline task/write-file/0.1/write-file.yaml:40]: Can we get the image tagged with digest as well? I mean in the format `docker.io/library/alpine:tag@digest`

Comment (bittrance) [reply task/write-file/0.1/write-file.yaml]: And I was going on about it on Slack too :facepalm: . Fixed now.

File: task/write-file/0.1/write-file.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  params:
+    - name: path
+      type: string
+      description: >
+        Relative path to create within the workspace. Directories will
+        be created as necessary.
+    - name: contents
+      type: string
+      description: >
+        Contents of the file to create. Note that octal numbers
```
Comment (sbwsg) [inline task/write-file/0.1/write-file.yaml:27]: Should this note on octal be in the `mode` param or is it intentionally here as part of `contents`?

**Comment by sbwsg** [comment]
> This seems to have provoked the bot to remove your lgtm label. Sorry about that.

All good, working as intended.

**Pull request merged**